TestFoundation/Utilities.swift

```diff
@@ -0,0 +1,105 @@
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+
+
+func checkHashableMutations_ValueType<Item: Hashable, S: Sequence>(
+    _ item: Item,
+    _ keyPath: WritableKeyPath<Item, S.Element>,
+    _ values: S,
+    file: StaticString = #file,
+    line: UInt = #line
+) {
+    _checkHashableMutations(
+        Item.self, Item.self,
+        item,
+        { $0 },
+        keyPath,
+        values)
+}
+
+func checkHashableMutations_NSCopying<Item: NSObject & NSCopying, S: Sequence>(
+    _ item: Item,
+    _ keyPath: ReferenceWritableKeyPath<Item, S.Element>,
+    _ values: S,
+    file: StaticString = #file,
+    line: UInt = #line
+) {
+    _checkHashableMutations(
+        Item.self, Item.self,
+        item,
+        { $0.copy() as! Item },
+        keyPath,
+        values)
+}
+
+func checkHashableMutations_NSMutableCopying<
+  Source: NSObject & NSMutableCopying,
+  Target: NSObject & NSMutableCopying,
+  S: Sequence
+>(
+    _ item: Source,
+    _ keyPath: ReferenceWritableKeyPath<Target, S.Element>,
+    _ values: S,
+    file: StaticString = #file,
+    line: UInt = #line
+) {
+    _checkHashableMutations(
+        Source.self, Target.self,
+        item,
+        { $0.mutableCopy() as! Target },
+        keyPath,
+        values)
+}
+
+// Check that mutating `object` via the specified key path affects its
+// hash value.
+func _checkHashableMutations<Source: Hashable, Target: Hashable, S: Sequence>(
```

Comment:
I know these are a ton of arguments and it's essentially fileprivate, but honestly I'd have a more descriptive set of labels because it is _really_ hard to parse the invocations above at a glance.

e.g.: `_checkHashing(of type, matches type, initialValue:, copyHandler:, byMutating keyPaths, throughRange:)`